### PR TITLE
Assessment flag for ML data training

### PR DESCRIPTION
**Pull request by rabstejnek**
- [x] Add boolean flag to hawc.apps.assessment.models.Assessment to allow sharing of data for machine learning projects
- [x] Add to the help-text that all data will be anonymized by assessemnt id and user id
- [x] Add a runpython in the migration to set that value to True for any assessment that is public; add to the help-text that all data will be anonymized by assessemnt id and user id
- [x] Create a dummy API endpoint that returns assessment-ids that have the is_public_training_data flag set to True

### CODE REVIEW

**Comment by rabstejnek** [comment]
The dummy API endpoint right now returns a JSON of assessments and their data for those that allow public training data.

**Comment by rabstejnek** [comment]
@shapiromatron I will be adding the real endpoint to this PR; I requested a review before I was given specifics on how to implement it. If you'd like to decline the current request, I will re-request when I'm done!

**Comment by rabstejnek** [comment]
@shapiromatron 
Ready for review

File: hawc/apps/assessment/api.py [hunk 1]
```diff
@@ -170,6 +178,61 @@ def public(self, request):
...
+            "animal_group__species__name": "species_name",
+            "animal_group__strain__name": "strain_name",
+            "system": None,
+            "animal_group__experiment__study__title": "study_title",
+            "animal_group__experiment__study__abstract": "study_abstract",
+            "animal_group__experiment__study__identifiers__database": "db",
+            "animal_group__experiment__study__identifiers__unique_id": "db_id",
+        }
+        queryset = (
+            Endpoint.objects.filter(assessment__is_public_training_data=True)
```
Comment (shapiromatron) [inline hawc/apps/assessment/api.py:197]: you should be able to view the django debug toolbar to see if this helps with the query, but you can also add a select_related_block

```python
select_related("animal_group", "animal_group__experiment", "animal_group__experiment__study")
```
```

File: hawc/apps/assessment/api.py [hunk 2]
```diff
@@ -170,6 +178,61 @@ def public(self, request):
...
+            """
+            hashed_id = hashlib.md5(str(id).encode())
+            hashed_id.update(settings.SECRET_KEY.encode())
+            return uuid.UUID(bytes=hashed_id.digest())
+
+        # Creates a UUID for each assessment_id, providing anonymity
+        df["assessment_uuid"] = df["assessment_uuid"].apply(create_uuid)
+
+        # Assigns db_id to hero_id in all instances where db == HERO
+        df["hero_id"] = None
```
Comment (shapiromatron) [inline hawc/apps/assessment/api.py:225]: nice job

File: hawc/apps/assessment/api.py [hunk 3]
```diff
@@ -170,6 +178,61 @@ def public(self, request):
...
+    @method_decorator(cache_page(60 * 60 * 24))
+    @action(detail=False, methods=("get",), renderer_classes=PandasRenderers)
+    def training_data(self, request):
+        Endpoint = apps.get_model("animal", "Endpoint")
+        # map of django field names to friendlier column names
+        column_map = {
+            "assessment_id": "assessment_uuid",
+            "animal_group__species__name": "species_name",
+            "animal_group__strain__name": "strain_name",
+            "system": None,
```
Comment (shapiromatron) [inline hawc/apps/assessment/api.py:190]: almost forgot; can we also add "organ", "effect", "effect_subtype", "data_location",  "name", "animal_group__sex",  "animal_group__experiment__chemical",

File: hawc/apps/assessment/api.py [hunk 4]
```diff
@@ -170,6 +175,60 @@ def public(self, request):
...
+            "animal_group__species__name": "species",
+            "animal_group__strain__name": "strain",
+            "animal_group__sex": "sex",
+            "system": None,
+            "organ": None,
+            "effect": None,
+            "effect_subtype": None,
+            "animal_group__experiment__study__title": "study_title",
+            "animal_group__experiment__study__abstract": "study_abstract",
+            "data_location": None,
```
Comment (shapiromatron) [inline hawc/apps/assessment/api.py:196]: let's rename `data_location` to `endpoint_data_location` to be clear what we're referring to

File: hawc/apps/assessment/api.py [hunk 5]
```diff
@@ -170,6 +175,60 @@ def public(self, request):
...
+            "name": None,
+            "animal_group__experiment__chemical": "chemical",
+            "animal_group__species__name": "species",
+            "animal_group__strain__name": "strain",
+            "animal_group__sex": "sex",
+            "system": None,
+            "organ": None,
+            "effect": None,
+            "effect_subtype": None,
+            "animal_group__experiment__study__title": "study_title",
```
Comment (shapiromatron) [inline hawc/apps/assessment/api.py:194]: can we also add `animal_group__experiment__study__full_citation`? that should give people what they need to pull

File: hawc/apps/assessment/api.py [hunk 6]
```diff
@@ -170,6 +175,60 @@ def public(self, request):
...
         return Response(serializer.data)
 
+    @method_decorator(cache_page(60 * 60 * 24))
+    @action(detail=False, methods=("get",), renderer_classes=PandasRenderers)
+    def training_data(self, request):
+        Endpoint = apps.get_model("animal", "Endpoint")
+        # map of django field names to friendlier column names
+        column_map = {
+            "assessment_id": "assessment_uuid",
+            "name": None,
```
Comment (shapiromatron) [inline hawc/apps/assessment/api.py:185]: let's prefix name, system, organ, effect, and effect_subtype with `endpoint`

File: hawc/apps/assessment/templates/assessment/assessment_detail.html [hunk 7]
```diff
@@ -138,6 +138,11 @@ <h1>{{object}}
           <td>{{object.public}}</td>
         </tr>
 
+        <tr>
+          <th>Public Training Data</th>
```
Comment (shapiromatron) [inline hawc/apps/assessment/templates/assessment/assessment_detail.html:142]: Change to `Public training data`; I know it's inconsistent in HAWC, but I try to capitalize the first word only

File: hawc/apps/common/helper.py [hunk 8]
```diff
@@ -62,6 +65,15 @@ def tryParseInt(val, default=None):
         return default
 
 
+def create_uuid(id: int) -> str:
```
Comment (shapiromatron) [inline hawc/apps/common/helper.py:68]: simple unit test for this?

File: hawc/apps/assessment/api.py [hunk 9]
```diff
@@ -1,17 +1,22 @@
 import logging
```
Comment (shapiromatron) [inline hawc/apps/assessment/api.py:1]: didn't know where else to put this; can we add to the client?

File: hawc/apps/assessment/api.py [hunk 10]
```diff
@@ -170,6 +175,60 @@ def public(self, request):
         serializer = serializers.AssessmentSerializer(queryset, many=True)
         return Response(serializer.data)
 
+    @method_decorator(cache_page(60 * 60 * 24))
+    @action(detail=False, methods=("get",), renderer_classes=PandasRenderers)
+    def training_data(self, request):
```
Comment (shapiromatron) [inline hawc/apps/assessment/api.py:180]: let's call this `bioassay-ml-dataset`, we'll likely have a different api call for epi

**Pull request merged**